File: 0330_patching_array/min_patches.py
from typing import List


def minPatches(nums: List[int], n: int) -> int:
    """
    Time: O(n)
    Space: O(n)
    """
    missing = 1
    i = 0
    patches = []
    while missing <= n:
        if i < len(nums) and nums[i] <= missing:
            missing += nums[i]
            i += 1
        else:
            patches.append(missing)
            missing += missing
    return len(patches)

File: 0330_patching_array/min_patches.java
class Solution {
    // Time: O(n)
    // Space: O(n)
    public int minPatches(int[] nums, int n) {
        long missing = 1;
        int i = 0;
        ArrayList<Long> patches = new ArrayList<>();
        while (missing <= n) {
            if (i < nums.length && nums[i] <= missing) {
                missing += nums[i++];
            } else {
                patches.add(missing);
                missing += missing;
            }
        }
        return patches.size();
    }
}